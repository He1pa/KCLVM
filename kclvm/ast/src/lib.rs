//! Copyright The KCL Authors. All rights reserved.

pub mod ast;
pub mod config;
pub mod path;
pub mod pos;
pub mod token;
pub mod token_stream;
pub mod walker;
use std::sync::Mutex;

use lazy_static::lazy_static;

#[cfg(test)]
mod tests;

pub const MAIN_PKG_DEFAULT: &str = "__main__";

lazy_static! {
    static ref MAIN_PKG: Mutex<String> = Mutex::new(MAIN_PKG_DEFAULT.to_owned());
}

pub fn set_main_pkg(package: &str) {
    let mut main_pkg = MAIN_PKG.lock().unwrap();
    *main_pkg = package.to_owned();
}

pub fn get_main_pkg() -> String {
    MAIN_PKG.lock().unwrap().clone()
}

#[macro_export]
macro_rules! node_ref {
    ($node: expr) => {
        NodeRef::new(Node::dummy_node($node))
    };
    ($node: expr, $pos:expr) => {
        NodeRef::new(Node::node_with_pos($node, $pos))
    };
}

#[macro_export]
macro_rules! expr_as {
    ($expr: expr, $expr_enum: path) => {
        if let $expr_enum(x) = ($expr.node as Expr) {
            Some(x)
        } else {
            None
        }
    };
}

#[macro_export]
macro_rules! stmt_as {
    ($stmt: expr, $stmt_enum: path) => {
        if let $stmt_enum(x) = ($stmt.node as Stmt) {
            Some(x)
        } else {
            None
        }
    };
}
