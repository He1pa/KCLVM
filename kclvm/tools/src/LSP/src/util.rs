use parking_lot::RwLockReadGuard;
use ra_ap_vfs::{FileId, Vfs};
use serde::{de::DeserializeOwned, Serialize};

/// Deserializes a `T` from a json value.
pub fn from_json<T: DeserializeOwned>(
    what: &'static str,
    json: serde_json::Value,
) -> anyhow::Result<T> {
    T::deserialize(&json)
        .map_err(|e| anyhow::anyhow!("could not deserialize {}: {}: {}", what, e, json))
}

/// Converts the `T` to a json value
pub fn to_json<T: Serialize>(value: T) -> anyhow::Result<serde_json::Value> {
    serde_json::to_value(value).map_err(|e| anyhow::anyhow!("could not serialize to json: {}", e))
}

pub fn get_file_name(vfs: RwLockReadGuard<Vfs>, file_id: FileId) -> anyhow::Result<String> {
    if let Some(path) = vfs.file_path(file_id).as_path() {
        Ok(path
            .as_ref()
            .to_str()
            .ok_or(anyhow::anyhow!("Failed to get file file"))?
            .to_string())
    } else {
        Err(anyhow::anyhow!(
            "{} isn't on the file system.",
            vfs.file_path(file_id)
        ))
    }
}
