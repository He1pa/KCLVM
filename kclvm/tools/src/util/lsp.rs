use kclvm_error::Diagnostic as KCLDiagnostic;
use kclvm_error::Level;
use kclvm_error::Message;
use kclvm_error::Position as KCLPos;
use tower_lsp::lsp_types::*;

use kclvm_sema::resolver::scope::Scope;
use kclvm_sema::resolver::scope::ScopeObject;

#[derive(Debug)]
pub struct ImCompleteSemanticToken {
    pub start: KCLPos,
    /// The scope object end position.
    pub end: KCLPos,
    /// The scope object kind.
    pub kind: u32,
    pub length: u32,
}

pub const LEGEND_TYPE: &[SemanticTokenType] = &[
    SemanticTokenType::FUNCTION,
    SemanticTokenType::VARIABLE,
    SemanticTokenType::STRING,
    SemanticTokenType::STRUCT,
    SemanticTokenType::COMMENT,
    SemanticTokenType::NUMBER,
    SemanticTokenType::KEYWORD,
    SemanticTokenType::OPERATOR,
    SemanticTokenType::PARAMETER,
    SemanticTokenType::PROPERTY,
];

/// Convert pos format
/// The position in lsp protocol is different with position in ast node whose line number is 1 based.
pub fn kcl_pos_to_lsp_pos(pos: &KCLPos) -> Position {
    Position {
        line: pos.line as u32 - 1,
        character: pos.column.unwrap_or(0) as u32,
    }
}

/// Convert KCL Message to LSP Diagnostic
fn kcl_msg_to_lsp_diags(msg: &Message, severity: DiagnosticSeverity) -> Diagnostic {
    let kcl_pos = msg.pos.clone();
    let start_position = kcl_pos_to_lsp_pos(&kcl_pos);
    let end_position = kcl_pos_to_lsp_pos(&kcl_pos);
    Diagnostic {
        range: Range::new(start_position, end_position),
        severity: Some(severity),
        code: None,
        code_description: None,
        source: None,
        message: msg.message.clone(),
        related_information: None,
        tags: None,
        data: None,
    }
}

fn kcl_err_level_to_severity(level: Level) -> DiagnosticSeverity {
    match level {
        Level::Error => DiagnosticSeverity::ERROR,
        Level::Warning => DiagnosticSeverity::WARNING,
        Level::Note => DiagnosticSeverity::HINT,
    }
}

/// Convert KCL Diagnostic to LSP Diagnostics.
/// Because the diagnostic of KCL contains multiple messages, and each messages corresponds to a diagnostic of LSP, the return value is a vec
pub fn kcl_diag_to_lsp_diags(diag: &KCLDiagnostic, file_name: &str) -> Vec<Diagnostic> {
    diag.messages
        .iter()
        .filter(|msg| msg.pos.filename == file_name)
        .map(|msg| kcl_msg_to_lsp_diags(msg, kcl_err_level_to_severity(diag.level)))
        .collect()
}

pub fn kcl_scopeobj_to_lsp_imcomplete_sema_token(
    obj: &ScopeObject,
) -> Option<ImCompleteSemanticToken> {
    match obj.kind {
        kclvm_sema::resolver::scope::ScopeObjectKind::Variable => Some(ImCompleteSemanticToken {
            start: obj.start.clone(),
            end: obj.end.clone(),
            kind: LEGEND_TYPE
                .iter()
                .position(|item| item == &SemanticTokenType::VARIABLE)
                .unwrap() as u32,
            length: obj.name.len() as u32,
        }),
        kclvm_sema::resolver::scope::ScopeObjectKind::Attribute => Some(ImCompleteSemanticToken {
            start: obj.start.clone(),
            end: obj.end.clone(),
            kind: LEGEND_TYPE
                .iter()
                .position(|item| item == &SemanticTokenType::PROPERTY)
                .unwrap() as u32,
            length: obj.name.len() as u32,
        }),
        kclvm_sema::resolver::scope::ScopeObjectKind::Definition => {
            let mut start_pos = obj.start.clone();
            start_pos.column =  Some(start_pos.column.unwrap_or(0) + "schema ".len() as u64  );
            Some(ImCompleteSemanticToken {
            start: start_pos,
            end: obj.end.clone(),
            kind: LEGEND_TYPE
                .iter()
                .position(|item| item == &SemanticTokenType::STRUCT)
                .unwrap() as u32,
            length: obj.name.len() as u32,
        })},
        kclvm_sema::resolver::scope::ScopeObjectKind::Parameter => Some(ImCompleteSemanticToken {
            start: obj.start.clone(),
            end: obj.end.clone(),
            kind: LEGEND_TYPE
                .iter()
                .position(|item| item == &SemanticTokenType::PARAMETER)
                .unwrap() as u32,
            length: obj.name.len() as u32,
        }),
        kclvm_sema::resolver::scope::ScopeObjectKind::TypeAlias => None,
        kclvm_sema::resolver::scope::ScopeObjectKind::Module => None,
    }
}

pub fn get_scope_imcomplete_sema_token(scope: &Scope, file_name: &str) -> Vec<ImCompleteSemanticToken>{
    let mut tokens = vec![];
    for (_, obj) in &scope.elems {
        let obj = obj.borrow();
        if obj.start.filename == file_name{
            if let Some(token) = kcl_scopeobj_to_lsp_imcomplete_sema_token(&obj){
                tokens.push(token);
            }
        }
    }
    for child in &scope.children{
        let child = child.borrow();
        tokens.append(&mut get_scope_imcomplete_sema_token(&child, file_name));
    }
    tokens
}
