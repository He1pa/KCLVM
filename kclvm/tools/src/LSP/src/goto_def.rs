use kclvm_ast::ast::{AssignStmt, Program, Stmt, UnificationStmt};
use kclvm_ast::pos::ContainsPos;
use kclvm_error::Position as KCLPos;
use kclvm_sema::resolver::scope::ProgramScope;
use lsp_types::{GotoDefinitionResponse, Url};
use lsp_types::{Location, Range};

use crate::to_lsp::lsp_pos;

pub(crate) fn goto_definition(
    program: Program,
    kcl_pos: KCLPos,
    prog_scope: ProgramScope,
) -> Option<lsp_types::GotoDefinitionResponse> {
    match program.pos_to_stmt(&kcl_pos) {
        Some(node) => match node.node {
            Stmt::Unification(stmt) => goto_def_for_unification(stmt, kcl_pos, prog_scope),
            Stmt::Assign(stmt) => goto_def_for_assign(stmt, kcl_pos, prog_scope),
            _ => None,
        },
        None => None,
    }
}

pub(crate) fn goto_def_for_unification(
    stmt: UnificationStmt,
    kcl_pos: KCLPos,
    prog_scope: ProgramScope,
) -> Option<GotoDefinitionResponse> {
    let schema_expr = stmt.value.node;
    if schema_expr.name.contains_pos(&kcl_pos)

    {
        let id = schema_expr.name.node.names.last().unwrap();
        let positions = find_name_in_program_scope(id, prog_scope);
        positions_to_goto_def_resp(&positions)
    } else {
        None
    }
}

pub(crate) fn goto_def_for_assign(
    stmt: AssignStmt,
    kcl_pos: KCLPos,
    prog_scope: ProgramScope,
) -> Option<GotoDefinitionResponse> {
    let id = {
        if let Some(ty) = stmt.type_annotation {
            if ty.contains_pos(&kcl_pos) {
                Some(ty.node)
            } else {
                None
            }
        } else if stmt.value.contains_pos(&kcl_pos){
            match stmt.value.node {
                kclvm_ast::ast::Expr::Identifier(id) => Some(id.names.last().unwrap().clone()),
                kclvm_ast::ast::Expr::Schema(schema_expr) => {
                    if schema_expr.name.contains_pos(&kcl_pos) {
                        Some(schema_expr.name.node.names.last().unwrap().clone())
                    } else {
                        None
                    }
                },
                _ => None

            }
        } else {
            None
        }
    };
    match id {
        Some(id) => {
            let positions = find_name_in_program_scope(&id, prog_scope);
            positions_to_goto_def_resp(&positions)
        }
        None => None,
    }
}

pub(crate) fn find_name_in_program_scope(
    name: &String,
    prog_scope: ProgramScope,
) -> Vec<(KCLPos, KCLPos)> {
    let mut positions = vec![];
    let mut scopes = vec![];
    for s in prog_scope.scope_map.values() {
        scopes.push(s.borrow().clone());
    }

    while !scopes.is_empty() {
        let s = scopes.pop().unwrap();
        match s.lookup(&name) {
            Some(obj) => {
                let obj = obj.borrow().clone();
                positions.push((obj.start, obj.end));
            }
            None => {
                for c in s.children {
                    scopes.push(c.borrow().clone());
                }
            }
        }
    }
    positions
}

pub(crate) fn positions_to_goto_def_resp(
    positions: &Vec<(KCLPos, KCLPos)>,
) -> Option<GotoDefinitionResponse> {
    match positions.len() {
        0 => None,
        1 => {
            let (start, end) = positions[0].clone();
            Some(lsp_types::GotoDefinitionResponse::Scalar(Location {
                uri: Url::from_file_path(start.filename.clone()).unwrap(),
                range: Range {
                    start: lsp_pos(&start),
                    end: lsp_pos(&end),
                },
            }))
        }
        _ => {
            let mut res = vec![];
            for (start, end) in positions {
                res.push(Location {
                    uri: Url::from_file_path(start.filename.clone()).unwrap(),
                    range: Range {
                        start: lsp_pos(&start),
                        end: lsp_pos(&end),
                    },
                })
            }
            Some(lsp_types::GotoDefinitionResponse::Array(res))
        }
    }
}
