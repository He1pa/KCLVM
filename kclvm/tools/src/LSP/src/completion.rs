//! Complete for KCL
//! Github Issue: https://github.com/kcl-lang/kcl/issues/476
//! Now supports code completion in treigger mode (triggered when user enters `.`),
//! and the content of the completion includes:
//!  + import path
//!  + schema attr
//!  + builtin function(str function)
//!  + defitions in pkg
//!  + system module functions

use std::io;
use std::{fs, path::Path};

use indexmap::IndexSet;
use kclvm_ast::ast::{Expr, ImportStmt, Program, Stmt};
use kclvm_compiler::pkgpath_without_prefix;
use kclvm_config::modfile::KCL_FILE_EXTENSION;

use kclvm_error::Position as KCLPos;
use kclvm_sema::builtin::{
    get_system_module_members, STANDARD_SYSTEM_MODULES, STRING_MEMBER_FUNCTIONS,
};
use kclvm_sema::resolver::scope::ProgramScope;
use lsp_types::CompletionItem;

use crate::goto_def::{get_identifier_last_name, resolve_var};
use crate::util::inner_most_expr_in_stmt;
use crate::util::{fix_missing_identifier, get_pkg_scope};

/// Computes completions at the given position.
pub(crate) fn completion(
    trigger_character: Option<char>,
    program: &Program,
    pos: &KCLPos,
    prog_scope: &ProgramScope,
) -> Option<lsp_types::CompletionResponse> {
    if let Some('.') = trigger_character {
        completion_dot(program, pos, prog_scope)
    } else {
        // todo: Complete identifiers such as attr, variables, types, etc.
        None
    }
}

fn completion_dot(
    program: &Program,
    pos: &KCLPos,
    prog_scope: &ProgramScope,
) -> Option<lsp_types::CompletionResponse> {
    // Get the position of trigger_character '.'
    let pos = &KCLPos {
        filename: pos.filename.clone(),
        line: pos.line,
        column: pos.column.map(|c| c - 1),
    };

    match program.pos_to_stmt(pos) {
        Some(node) => match node.node {
            Stmt::Import(stmt) => completion_for_import(&stmt, pos, prog_scope, program),
            _ => {
                let expr = inner_most_expr_in_stmt(&node.node, pos, None).0;
                match expr {
                    Some(node) => {
                        let items = get_completion_items(&node.node, prog_scope);
                        Some(into_completion_items(&items).into())
                    }
                    None => None,
                }
            }
        },
        None => None,
    }
}

fn completion_for_import(
    stmt: &ImportStmt,
    _pos: &KCLPos,
    _prog_scope: &ProgramScope,
    program: &Program,
) -> Option<lsp_types::CompletionResponse> {
    let mut items: IndexSet<String> = IndexSet::new();
    let pkgpath = &stmt.path;
    let real_path =
        Path::new(&program.root).join(pkgpath.replace('.', &std::path::MAIN_SEPARATOR.to_string()));
    if real_path.is_dir() {
        if let Ok(entries) = fs::read_dir(real_path) {
            let mut entries = entries
                .map(|res| res.map(|e| e.path()))
                .collect::<Result<Vec<_>, io::Error>>()
                .unwrap();
            entries.sort();
            for path in entries {
                let filename = path.file_name().unwrap().to_str().unwrap().to_string();
                if path.is_dir() {
                    items.insert(filename);
                } else if path.is_file() {
                    if let Some(extension) = path.extension() {
                        if extension == KCL_FILE_EXTENSION {
                            items.insert(
                                path.with_extension("")
                                    .file_name()
                                    .unwrap()
                                    .to_str()
                                    .unwrap()
                                    .to_string(),
                            );
                        }
                    }
                }
            }
        }
    }
    Some(into_completion_items(&items).into())
}

fn get_completion_items(expr: &Expr, prog_scope: &ProgramScope) -> IndexSet<String> {
    let mut items = IndexSet::new();
    match expr {
        Expr::Identifier(id) => {
            let name = get_identifier_last_name(&id);
            if !id.pkgpath.is_empty() {
                // standard system module
                if STANDARD_SYSTEM_MODULES.contains(&name.as_str()) {
                    items.extend(
                        get_system_module_members(name.as_str())
                            .iter()
                            .map(|s| s.to_string()),
                    )
                }
                // user module
                if let Some(scope) = prog_scope
                    .scope_map
                    .get(&pkgpath_without_prefix!(id.pkgpath))
                {
                    let scope = scope.borrow();
                    for (name, obj) in &scope.elems {
                        if obj.borrow().ty.is_module() {
                            continue;
                        }
                        items.insert(name.clone());
                    }
                }
                return items;
            }

            let def = resolve_var(
                &fix_missing_identifier(&id.names),
                &get_pkg_scope(&id.pkgpath, &prog_scope.scope_map),
                &prog_scope.scope_map,
            );

            if let Some(def) = def {
                match def {
                    crate::goto_def::Definition::Object(obj) => {
                        match &obj.ty.kind {
                            // builtin (str) functions
                            kclvm_sema::ty::TypeKind::Str => {
                                let binding = STRING_MEMBER_FUNCTIONS;
                                for k in binding.keys() {
                                    items.insert(format!("{}{}", k, "()"));
                                }
                            }
                            // schema attrs
                            kclvm_sema::ty::TypeKind::Schema(schema) => {
                                for k in schema.attrs.keys() {
                                    if k != "__settings__" {
                                        items.insert(k.clone());
                                    }
                                }
                            }
                            _ => {}
                        }
                    }
                    crate::goto_def::Definition::Scope(_) => {}
                }
            }
        }
        Expr::Selector(select_expr) => {
            let res = get_completion_items(&select_expr.value.node, prog_scope);
            items.extend(res);
        }
        Expr::StringLit(_) => {
            let binding = STRING_MEMBER_FUNCTIONS;
            for k in binding.keys() {
                items.insert(format!("{}{}", k, "()"));
            }
        }
        _ => {}
    }
    items
}

pub(crate) fn into_completion_items(items: &IndexSet<String>) -> Vec<CompletionItem> {
    items
        .iter()
        .map(|item| CompletionItem {
            label: item.clone(),
            ..Default::default()
        })
        .collect()
}
