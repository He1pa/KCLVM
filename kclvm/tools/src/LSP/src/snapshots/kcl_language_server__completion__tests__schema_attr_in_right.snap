---
source: tools/src/LSP/src/completion.rs
expression: "format!(\"{:?}\", got_labels)"
---
["Base", "Base{}", "Config", "Config{}", "Name", "Name{}", "Person", "Person(b){}", "n", "p"]
